core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala

```
@@ -109,73 +109,91 @@ class TransactionCoordinator(brokerId: Int,
+            epochAndMetadata synchronized {
+              if (epochAndMetadata.transactionMetadata.pendingTransitionInProgress)
+                responseCallback(initTransactionError(Errors.CONCURRENT_TRANSACTIONS))
+              else if (!epochAndMetadata.transactionMetadata.equals(newMetadata))
+                initPidWithExistingMetadata(transactionalId, transactionTimeoutMs, responseCallback, epochAndMetadata)
               else
-                appendMetadataToLog(transactionalId, metadata, responseCallback)
-
+                initPidWithNewMetadata(transactionalId, responseCallback, epochAndMetadata)
+            }
+          case Some(existingEpochAndMetadata) =>
+            existingEpochAndMetadata synchronized {
+              initPidWithExistingMetadata(transactionalId, transactionTimeoutMs, responseCallback, existingEpochAndMetadata)
             }
-          case Some(metadata) =>
-            initPidWithExistingMetadata(transactionalId, transactionTimeoutMs, responseCallback, metadata)
         }
       }
   }
 
-  private def appendMetadataToLog(transactionalId: String,
-                             metadata: TransactionMetadata,
-                             initPidCallback: InitPidCallback): Unit ={
+  private def initPidWithNewMetadata(transactionalId: String,
+                                     initPidCallback: InitPidCallback,
+                                     epochAndMetadata: CoordinatorEpochAndTxnMetadata): Unit = {
+    val metadata = epochAndMetadata.transactionMetadata
+
     def callback(errors: Errors): Unit = {
       if (errors == Errors.NONE)
         initPidCallback(initTransactionMetadata(metadata))
       else
         initPidCallback(initTransactionError(errors))
     }
-    txnManager.appendTransactionToLog(transactionalId, metadata, callback)
+
+    // set the pending state to Empty again to allow the metadata to be updated
+    // after it was successfully written to log
+    metadata.prepareTransitionTo(Empty)
```

This is a comment throughout, but it seems we rarely check the result of `prepareTransitionTo`. Perhaps that function should just raise an invalid state exception if the state transition is invalid? Otherwise we should get in the habit of adding the checks.

---

We check it in `TransactionMetadata`, that the new state is equal to the pending state.

---

But if we know the transition is invalid here, then we could skip appending to the log, right? I feel we should just make `prepareTransitionTo` raise an exception if the attempted transition is invalid.

---

EDIT: discussed offline, realized it is a different issue than I thought about. Refactored this part a bit as well.